feat: notifications - authentication controller

Prithpal-Sooriya opened the pull request:
## **Description**
Pre-requisite PR that adds an authentication controller to enable API calls to MetaMask authorised endpoints (e.g. notifications, profile syncing, etc)

Closes: https://consensyssoftware.atlassian.net/browse/NOTIFY-414?atlOrigin=eyJpIjoiZGQ4MTMxMGI0ZTk3NGVmNmE1YTkzOWQ0M2ZkYWNkMGEiLCJwIjoiaiJ9

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23286?quickstart=1)

**NOTE** - this Controller is not connected to anywhere in the UI. Following work items will integrate it.

## **Manual testing steps**

There are no manual steps. This controller is not connected to anything (yet)

## **Screenshots/Recordings**

N/A


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Prithpal-Sooriya committed:
feat: notifications - authentication controller

adds the initial authentication controller to enable calls to authorized endpoints

Prithpal-Sooriya commented:
I have read the CLA Document and I hereby sign the CLA

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
ci: add notification auth endpoints and info to build

adds the dev auth server endpoints and info.
We can move this into github variables, but no/little risk to expose this on client.

Prithpal-Sooriya commented on app/scripts/metamask-controller.js:
```diff
@@ -299,6 +299,10 @@ import { snapKeyringBuilder, getAccountsBySnapId } from './lib/snap-keyring';
 import { addDappTransaction, addTransaction } from './lib/transaction/util';
 import { LatticeKeyringOffscreen } from './lib/offscreen-bridge/lattice-offscreen-keyring';
+///: BEGIN:ONLY_INCLUDE_IF(notifications)
+import { PREINSTALLED_SNAPS } from './snaps/preinstalled-snaps';
```
Technically our notifications feature if also tied to preinstalled snaps.

Lets move the preinstalled snaps to the correct code fence.

Prithpal-Sooriya committed:
refactor: move preinstalled snaps into snaps codefence

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
refactor: fix ci failures

fixes duplicate deps, linting errors, and mocha ESM script failures

Prithpal-Sooriya committed:
refactor: fix lavamoat build issues

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
refactor: remove notifications for some builds

Prithpal-Sooriya committed:
test: test cleanup

use jest.tothrow for cleaner assertion tests

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
chore: update policy

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

DDDDDanica commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,32 @@
+const snapId = 'npm:@metamask/message-signing-snap' as SnapId;
+
+export function createSnapPublicKeyRequest() {
+  return {
+    snapId,
+    origin: '',
+    handler: HandlerType.OnRpcRequest,
+    request: {
+      jsonrpc: '2.0',
+      method: 'getPublicKey',
+      params: [],
+      id: 1,
```
Can you extract return value as a type and specify it in the function signature ?

Prithpal-Sooriya replied:
Yep sure! The SnapController didn't have any nice single internal types. But can recreate or use it.
```
SnapRpcHookArgs & { snapId: SnapId; }
```

DDDDDanica commented on app/scripts/controllers/authentication/mocks/mockServices.ts:
```diff
@@ -0,0 +1,60 @@
+  OIDC_TOKENS_ENDPOINT,
+} from '../services';
+
+type MockReply = {
+  status: number;
+  body?: Record<string, unknown>;
+};
```
Can we leverage `express.Response` as the type for `mockReply`?

Prithpal-Sooriya replied:
This is a simple type used for the `nock` functions below.

I've had a look at express and seems a bit overkill.
1. The Response type is too wide for this simple use case (but we can always `Pick<T>` out the specifics we need.)
2. The Body is too loose being an `any`

It might be fine to keep the type as is or tie it to nock :)

danjm commented on app/scripts/controllers/authentication/authentication-controller.ts:
```diff
@@ -0,0 +1,247 @@
+import {
```
Was there consideration given to adding these controllers as their own package in https://github.com/MetaMask/core?

Prithpal-Sooriya replied:
Discussed, we'll keep this within extension & will move out once we start mobile notifications :)

DDDDDanica commented on app/scripts/metamask-controller.js:
```diff
@@ -1339,6 +1346,17 @@ export default class MetamaskController extends EventEmitter {
 
+    ///: BEGIN:ONLY_INCLUDE_IF(notifications)
+    // Notification Controllers
+    this.authenticationController = new AuthenticationController({
```
Not a big deal question: why is the controller named as `authenticationController` instead of `notificationController` ?

Prithpal-Sooriya replied:
Great Q.

We will have a set of controllers that the notifications flag will contain - each with a set of responsibilities.
This is an initial PR with just this controller (was recommended to split into multiple PRs to reduce PR size).

Prithpal-Sooriya committed:
Revert "chore: update policy"

This reverts commit d2e3179e1667af1e357ea8b5c10537f994ca4d94.

Prithpal-Sooriya commented on app/scripts/snaps/preinstalled-snaps.js:
```diff
@@ -0,0 +1,45 @@
+// eslint-disable-next-line import/unambiguous
```
I did attempt to convert this to a typescript file, but there were some issues around module resolution and require.resolve syntax - most likely ESM vs CJS syntax 🙃 

I'll see if I can re-try this. 👍🏾

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
chore: requested changes

Prithpal-Sooriya committed:
refactor: fix linting

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
build: use auth prod endpoints

Prithpal-Sooriya committed:
build: use prod auth client id

Prithpal-Sooriya commented:
Fixing breaking builds & CI. Will mark ready for review one done

Prithpal-Sooriya committed:
fix: failing builds due to importing fs directly

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

pedronfigueiredo commented on builds.yml:
```diff
@@ -145,6 +147,12 @@ features:
   conf-redesign:
+  notifications:
+    env:
+      - AUTH_API: https://authentication.api.cx.metamask.io
```
Is there a reason these are set only for the main and flask build types, not for all of them?

Prithpal-Sooriya replied:
These are our main targets. However not us against supporting all targets.

Prithpal-Sooriya replied:
I've discussed with the team. We'll initially support notifications for Main and Flask.
If the extension team prefers we support all, then that can be done (not a hard requirement)

Prithpal-Sooriya commented on app/scripts/controllers/authentication/authentication-controller.ts:
```diff
@@ -0,0 +1,247 @@
+        accessToken,
+      };
+    } catch (e) {
+      throw new Error(`${controllerName}: Failed to authenticate`);
```
Good find on error propagation. Yep I'll append the `controllerName` to error messages thrown from public methods.

Prithpal-Sooriya replied:
Do we support [Error: Cause](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) data property? Looks like our TS target supports it, I'll try it out.

Prithpal-Sooriya committed:
chore: requested changes

FrederikBolding commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,37 @@
+
+const snapId = 'npm:@metamask/message-signing-snap' as SnapId;
+
+export function createSnapPublicKeyRequest(): SnapRPCRequest {
```
Nit: Seems a bit unnecessary to have utility functions for these 😅

Prithpal-Sooriya replied:
Overkill yes. But these will be reused in future controllers 😃 (future PRs available soon).

FrederikBolding commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,37 @@
+      jsonrpc: '2.0',
+      method: 'getPublicKey',
+      params: [],
+      id: 1,
```
```suggestion
```

FrederikBolding commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,37 @@
+    origin: '',
+    handler: HandlerType.OnRpcRequest,
+    request: {
+      jsonrpc: '2.0',
```
```suggestion
```

FrederikBolding commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,37 @@
+    request: {
+      jsonrpc: '2.0',
+      method: 'getPublicKey',
+      params: [],
```
```suggestion
```

Prithpal-Sooriya replied:
love the defaults!!

FrederikBolding commented on app/scripts/controllers/authentication/auth-snap-requests.ts:
```diff
@@ -0,0 +1,37 @@
+    origin: '',
+    handler: HandlerType.OnRpcRequest,
+    request: {
+      jsonrpc: '2.0',
```
```suggestion
```

pedronfigueiredo commented on app/scripts/controllers/authentication/authentication-controller.ts:
```diff
@@ -0,0 +1,247 @@
+        accessToken,
+      };
+    } catch (e) {
+      throw new Error(`${controllerName}: Failed to authenticate - ${e}`);
```
I haven't tested it but I think we would see [object Object] here. Perhaps we can JSON.stringify the error or pass the message only?

Prithpal-Sooriya replied:
Great point. Tested this.

Since we are using Errors, they will automatically use the `.toString()` when using string concatenation.

But just for safety, I'll add a guard in case if it isn't an `instanceof Error`.

```
const errorMessage = e instanceof Error ? e.toString() : JSON.stringify(e)
```

Prithpal-Sooriya commented on app/scripts/snaps/preinstalled-snaps.ts:
```diff
@@ -0,0 +1,49 @@
+const TEMP_PREINSTALLED_SNAPS = [];
+
+///: BEGIN:ONLY_INCLUDE_IF(snaps)
+try {
```
This is valid.
I just remember having issues when converting this file to TS and left it there.

But I think it works out well if we support additional pre-installed snaps & can fallback to the empty exported snaps above.

I might add some error logs in the catch :)

brad-decker replied:
@FrederikBolding do you find this answer satisfying?

FrederikBolding replied:
Ideally if something is wrong with this configuration we fail hard IMO. I would vote for removing the try-catch 😅

matteoscurati reviewed (dismissed):
Tested locally, nothing to report. In the future, we might abstract the controller, but for now, it's faster to integrate it as proposed!

Prithpal-Sooriya committed:
refactor: requested changes

small snaps changes and auth error logging

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
feat: add server session profile and metametrics id

this was supposed to go into a separate PR, but was small enough to keep in here.

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
test: update snapshots

The snaps registry snapshots needed updating

DDDDDanica commented on app/scripts/controllers/authentication/services.ts:
```diff
@@ -0,0 +1,96 @@
+      return r.ok ? r.json() : null;
+    })
+    .then((r: NonceResponse | null) => r?.nonce)
+    .catch(() => null);
```
should it be error catching here? if we don't catch, IMO there's no need for return null

Prithpal-Sooriya replied:
I'm not a fan of error propagation in JS.
The consumer of this function will then need to handle: string return; a null return; and a hidden potentially thrown error.

I'd rather catch as early as possible and either treat the result as a value (in this case `null`) or make it into an "option" type (e.g. `{ isValid: false, error: e }`).

In this case I'm happy to opt for resolving catches to `null` as there is not information I care about propagating up to the consumer (but might change in future).

brad-decker replied:
More to the point, if we are using the 'await' terminology here why are we using .then and .catch chaining? 

Is it to avoid intermediary naming? 

Aka
```
let nonce = null;
try {
  const nonceResponse = await fetch(nonceUrl.toString());
  nonce = nonceResponse.ok ? r.json()?.nonce : null;
} catch () {
  // possible logLevel or some sort of sentryMessage so that we're aware this is happening?
}
```

OR drop the await.

brad-decker replied:
> Do we not allow mixing of promises and await syntaxes?
(personal) I prefer .then chaining to keep sections (e.g. transformations) scoped. Likewise it allows error handling immediately than a global try-catch where you need to differentiate errors.

Mixing these styles is confusing, personally, to me, and we should be clear here that the goal of pull request reviews is to get code into a state where someone who isn't you can maintain that code afterward. I'd love to hear other opinions here and it isn't just my view versus yours its a collective. 

> (personal) I prefer .then chaining to keep sections (e.g. transformations) scoped.
What do you mean here by "scoped transformations"? 

> Likewise it allows error handling immediately than a global try-catch where you need to differentiate errors.

I'm not sure what you mean by a global try catch. The try catch in my example is scoped around the same code that the chain method covers.

Prithpal-Sooriya replied:
Updated comment see snippet below.

These are all personal coding decisions, and yep we can leave it to a collective. I however don't think we should dissuade patterns that may be handy (such as mixing async/await with promises) otherwise we might as well not adopt any modern syntax (any TS syntax or shorthands, what about future JS syntax like the pipe operator?).

If there are patterns we as a collective do agree not to use (e.g. Typescript Constructor Shorthand), then lets use concrete this through ESLint rules, otherwise keep it flexible if there are not logical issues.

I do agree that the above code was pretty terse (especially the `.then(r => r.ok ? null : r.json())`), and have opted for an async/await.

```ts
// Treat errors as values.
// Think similar to an Option type, Result type, or even a null error result.
const res1: Foo | null = await foo().then(/* do stuff*/).catch(() => null)
if (!res1) {
  // handle case here
}

const res2: Bar | null = await bar().then(/* do stuff*/).catch(() => null)
if (!res2) {
  // handle case here
}

// VS try-catch clauses
try {
  const res1: Foo = await foo()
  const res2: Bar = await bar()
} catch (e: unknown) {
  // error here is unknown, we need to somehow instanceof or "figure out" which error is which
}

// OR let-try-catch
// this is okay, but why have a mutable variable that is only set once?
let res1: Foo | undefined;
try {
  res1 = await foo()
} catch (e: unknown) {
  // handle error
}

let res2: Bar | undefined;
try {
  res2 = await bar()
} catch (e: unknown) {
  // handle error
}
```

DDDDDanica replied:
> I'm not a fan of hidden error propagation in JS/TS.

I believe there's a misunderstanding there. IMO return `null` does not help with the error catching, we should throw error out when the network request is failed. 

```
try {
  const res1: Foo = await foo()
  const res2: Bar = await bar()
} catch (e: unknown) {
  // error here is unknown, we need to somehow instanceof or "figure out" which error is which
}
```
I personally found this solution acceptable, as we actually don't want to fire 2nd request when 1st one is failed, and the error will be caught clearly. However, the `e: unknown` is a question, i believe the `foo()` will return the `error` object as we expect.

Prithpal-Sooriya replied:
> IMO return null does not help with the error catching
This is valid, but again personal. IMO if I wanted to handle the error I would usually "return" some error instead of throwing. See how these [services are used](https://github.com/MetaMask/metamask-extension/pull/23286/files#diff-2061685b6ae9e88df78ce92c7c0a5fcdaadf5f4b7fd09f8ce8e66aab6eb62c38R172) , we handle null's here and then can throw.

> However, the e: unknown is a question, i believe the foo() will return the error object as we expect
Close, TS actually infers the catch as an `unknown` - [see playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKBDAXFAzsAJwEsA7AcygF8AoDXEEgYygDMBXJ4IuE1hACgCUOAAoE4AWyK4IAHnhwAfGmpQowABbiA7lBIRdAUQLiC-AOQsBgqAFooEuIwDWUCCbgFzg6jVr0mVg5GLh51CHwhFTVCEGi1KEYefCgCCJwFZExtDCJgPjghVSpEjGBGDSh+CBtUYrUAegaEtQA9AH5qeqgmqAAjCEYMNhk3DwIoDXdoaSgOZxI4bRIAGihtaAg8qYn9CAATdURQSExcGQJ8vKhPOdHSfAwmCDgWRKmXXAA6bqI36qgD2AT0YLzexlMtW6aiSJBS7lMADkMB4NocUBAANzQnrNBIdbp+Fq9ACSEgwZFI0BAcDYE1i-QANk5XBoMIdySQ4vs-ixpiR8nQGMwhozGd9uqS3jS2OsnsADkcHE8KdBNNAeSw+WkBWNTKUZLg1jL1rTGYc9odgIg6Bd8rcKoNXBAMBUDRAfsTmgBBMi5XiaWazFhUgCEUBE7lwPAwjJuYFI3F4swAViN8mySPtGWqpnrbnRAfkkhIIqkyjt1Jn1AQQA0huUNJ6qL4gA).
You will either need to type assert it to the error or use an `instanceof` checks.

Again this is fine, and we are bike shedding since any solution works. Everyone has their own preference:
- Throw errors vs return errors
- Handle errors inline vs handle all errors on catch statement.

I've updated the code to use async await, but am still not throwing as there is no point.

brad-decker replied:
@Prithpal-Sooriya everyone making their own personal coding decisions in terms of patterns will lead us down a path of chaos, which is why we are having these conversations -- anything that is not understood by the code reviewers is a clear indicator of an issue with maintainability going forward. We are pushing towards standards and guidelines so that everyone's code can mesh effectively and so that the next person after you can effectively maintain it. That isn't bike shedding. These are important questions for the health of the code base.

Prithpal-Sooriya replied:
I like playing devils advocate here 😄 

Yep I **_mostly_** agree, but until these rules are concrete & we add linting rules, code should be flexible.
If code is documented, tested, and functions are short & clear I see no reason for these to be blockers.

The original/main comment:
> should it be error catching here? if we don't catch, **IMO** there's no need for return null

Is a discussion about error catching and if we should return null. We are all writing JavaScript & Typescript, there are probably many different error handling patterns you can create (now and in the future). Key thing here is that this is all opinions, we can even raise this see what others consider.

See the function definition, the return type, the tests and where/how it is used - I think it is completely fine to:
1. Catch any errors to prevent any unhandled errors propagating up.
2. Return null if the endpoint fails.

Is it fine to return null? I think so! 
Lets look at this function definition `getNonce(): Promise<string | null>`. As a consumer of this, I can see that this will may give me a nonce, or null. If it is a null, I can choose what I want to do (now throw an error, or even go through a different flow).

DDDDDanica commented on app/scripts/controllers/authentication/services.ts:
```diff
@@ -0,0 +1,96 @@
+      return r.ok ? r.json() : null;
+    })
+    .then((r: LoginResponse | null) => r)
+    .catch(() => null);
```
same as https://github.com/MetaMask/metamask-extension/pull/23286/files#r1523424705, better to handle error or drop the catch

Prithpal-Sooriya replied:
See related comment.

brad-decker replied:
See the comment in the other thread

DDDDDanica commented on app/scripts/controllers/authentication/services.ts:
```diff
@@ -0,0 +1,96 @@
+  return accessToken ?? null;
+}
+
+export function createLoginRawMessage(nonce: string, publicKey: string) {
```
```suggestion
export function createLoginRawMessage(nonce: string, publicKey: string): string {
```

brad-decker replied:
As of right now this isn't a hard requirement, but when we adopt the full ruleset of our eslint config it will be. https://github.com/MetaMask/eslint-config/blob/main/packages/typescript/src/index.js#L46C1-L47C1 

Did this conversation get resolved outside of git @DDDDDanica @Prithpal-Sooriya ? If so it should be followed up with in the thread for reviewer context. If it did not get followed up on it should not be marked resolved until it is addressed or the reviewer resolves it @Prithpal-Sooriya

DDDDDanica commented on app/scripts/snaps/preinstalled-snaps.ts:
```diff
@@ -0,0 +1,57 @@
+}
+///: END:ONLY_INCLUDE_IF
+
+export const PREINSTALLED_SNAPS = Object.freeze(TEMP_PREINSTALLED_SNAPS);
```
Question here: if we require `PREINSTALLED_SNAPS`, it is possible to be `[]` first and then the file processing happens, and then we have filled `TEMP_PREINSTALLED_SNAPS` with push. So correct me if i'm wrong, this racing condition could be an issue. Can we try this patter: export a function as `getPreinstalledSnaps` and use async/await to return the value instead?

Prithpal-Sooriya replied:
Unlike ESM `import()` syntax, `require` is synchronous, so is the `fs.readFileSync`.

Let treat this as a nit - we can refactor to make into an async function once we add more pre-installed snaps.

brad-decker replied:
@DDDDDanica did you find @Prithpal-Sooriya 's answer to this question satisfactory?

DDDDDanica replied:
@Prithpal-Sooriya can we have a ticket to track this refactor todo in order to not lose track of it? It is also helpful to add this todo ticket number in the codebase as well !

and @brad-decker yes, this is acceptable ~

FrederikBolding replied:
Preinstalled snaps are currently not asynchronous. The Snap source files need to be available when the `MetaMaskController` boots synchronously. The code that @Prithpal-Sooriya is proposing will mostly run at build-time to inline these required files in the build so that they can be used at runtime synchrnously.

I don't believe there is a race condition here.

Prithpal-Sooriya commented on test/e2e/tests/state-snapshots/errors-after-init-opt-in-background-state.json:
```diff
@@ -205,7 +205,7 @@
   "SnapsRegistry": {
     "database": "object",
-    "lastUpdated": "object",
+    "lastUpdated": "number",
```
OK I've had a look, this snapshot change seems valid.
Looking at the SnapRegistry this should be a number:
https://github.com/MetaMask/snaps/blob/main/packages/snaps-controllers/src/snaps/registry/json.ts#L86

I've had a look in the codebase, this field also is not used anywhere - so has low impact.

How this change occurred: I believe by installing the snap above also triggered an update to the `@metamask/snaps-controllers` package.

FrederikBolding replied:
This is actually unintended, and a symptom of a potential problem. This essentially means that during testing the Snaps registry was updated, but we only do that if users have Snaps installed to check if they need to be blocked. We don't need to do this for preinstalled Snaps. I will create a PR to get this fixed.

FrederikBolding replied:
https://github.com/MetaMask/metamask-extension/pull/23485

Prithpal-Sooriya replied:
OK nice find, I didn't know of this. Updated the branch and reran snapshots - all back to normal 👍🏾

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

brad-decker commented on app/scripts/snaps/preinstalled-snaps.ts:
```diff
@@ -0,0 +1,57 @@
+// const require = createRequire(import.meta.url)
+//                                ^ The 'import.meta' meta-property is not allowed in files which will build into CommonJS output
+/* eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires */
+const fs = require('fs');
```
Why are you overriding our eslint rules?

Prithpal-Sooriya replied:
See the comment above.

1. Importing `fs` or `node:fs` does not work due to (I believe) lavamoat.
2. Unfortunately I was unable to get the ESM `createRequire` syntax to work.

Options:
1. We convert this file to (C)JS so we can use the `require` syntax, but believe we are moving away from JS scripts.
2. We disable the linting rule.

brad-decker replied:
@davidmurdoch and or @legobeat do you have any feedback on this item?

just to note the fs that will be imported here is the browserify version of fs, not node. At least thats my understanding because this code is run in the browser which doesn't have fs.

Prithpal-Sooriya replied:
Oooh browserify, that must mean we are also using browserify's `require` here too.
If this linting rule is valid, then why did we not get any errors or warning when using `require.resolve`?

<details><summary>Summary when using  `import fs from 'fs'`</summary>

```
scripts:core:dev:standardEntryPoints: Error: tried to statically call { readFileSync: [Function: readFileSync], readFile: [Function: readFile], readdirSync: [Function: readdirSync], readdir: [Function: readdir] } as a function while parsing file: /Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/app/scripts/snaps/preinstalled-snaps.ts
    at error (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/static-module/index.js:167:45)
    at traverse (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/static-module/index.js:275:24)
    at walk (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/static-module/index.js:259:13)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:79:24)
    at Object.skipThrough (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:186:39)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at base.NewExpression.base.CallExpression (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:395:11)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at Object.skipThrough (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:186:39)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at base.VariableDeclarator (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:285:22)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at base.VariableDeclaration (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:280:7)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at Object.skipThrough (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:186:39)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at base.Program.base.BlockStatement (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:198:7)
    at c (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:78:24)
    at Object.full (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-walk/dist/walk.js:80:7)
    at full (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/acorn-node/walk.js:20:15)
    at /Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/static-module/index.js:57:13
    at ConcatStream.<anonymous> (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/concat-stream/index.js:37:43)
    at ConcatStream.emit (node:events:526:35)
    at finishMaybe (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:630:14)
    at endWritable (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:638:3)
    at Writable.end (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:594:41)
    at DuplexWrapper.<anonymous> (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/duplexer2/index.js:29:14)
    at Object.onceWrapper (node:events:628:28)
    at DuplexWrapper.emit (node:events:526:35)
    at finishMaybe (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:630:14)
    at endWritable (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:638:3)
    at Writable.end (/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:594:41)
    at BabelifyStream.onend (node:internal/streams/readable:924:10)
    at Object.onceWrapper (node:events:628:28)
    at BabelifyStream.emit (node:events:514:28)
    at endReadableNT (node:internal/streams/readable:1589:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  filename: '/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/app/scripts/snaps/preinstalled-snaps.ts',
  stream: Labeled {
    _readableState: ReadableState {
      objectMode: true,
      highWaterMark: 16,
      buffer: BufferList { length: 0 },
      length: 0,
      pipes: [Labeled],
      pipesCount: 1,
      flowing: true,
      ended: false,
      endEmitted: false,
      reading: true,
      sync: false,
      needReadable: true,
      emittedReadable: false,
      readableListening: false,
      resumeScheduled: false,
      destroyed: false,
      defaultEncoding: 'utf8',
      awaitDrain: 0,
      readingMore: false,
      decoder: null,
      encoding: null
    },
    readable: true,
    _events: [Object: null prototype] {
      end: [Array],
      error: [Function (anonymous)],
      data: [Function: ondata],
      _mutate: [Function]
    },
    _eventsCount: 4,
    _maxListeners: undefined,
    _writableState: WritableState {
      objectMode: true,
      highWaterMark: 16,
      finalCalled: false,
      needDrain: false,
      ending: true,
      ended: true,
      finished: true,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: false,
      bufferProcessing: false,
      onwrite: [Function (anonymous)],
      writecb: null,
      writelen: 0,
      bufferedRequest: null,
      lastBufferedRequest: null,
      pendingcb: 0,
      prefinished: true,
      errorEmitted: false,
      bufferedRequestCount: 0,
      corkedRequestsFree: [CorkedRequest]
    },
    writable: false,
    allowHalfOpen: true,
    _options: { objectMode: true },
    _wrapOptions: { objectMode: true },
    _streams: [ [Deps], [Transform] ],
    length: 2,
    label: 'deps',
    [Symbol(kCapture)]: false
  }
}
```
</details> 


<details><summary>Summary when using  `import fs from 'node:fs'`</summary>

```
scripts:core:dev:standardEntryPoints: [Error: Can't walk dependency graph: ENOENT: no such file or directory, lstat '/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node:fs'
    required by /Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/app/scripts/snaps/preinstalled-snaps.ts] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'lstat',
  path: '/Users/prithpalsooriya/Desktop/projects/metamask/metamask-extension/node:fs',
  stream: Labeled {
    _readableState: ReadableState {
      objectMode: true,
      highWaterMark: 16,
      buffer: BufferList { length: 0 },
      length: 0,
      pipes: [Labeled],
      pipesCount: 1,
      flowing: true,
      ended: false,
      endEmitted: false,
      reading: true,
      sync: false,
      needReadable: true,
      emittedReadable: false,
      readableListening: false,
      resumeScheduled: false,
      destroyed: false,
      defaultEncoding: 'utf8',
      awaitDrain: 0,
      readingMore: false,
      decoder: null,
      encoding: null
    },
    readable: true,
    _events: [Object: null prototype] {
      end: [Array],
      error: [Function (anonymous)],
      data: [Function: ondata],
      _mutate: [Function]
    },
    _eventsCount: 4,
    _maxListeners: undefined,
    _writableState: WritableState {
      objectMode: true,
      highWaterMark: 16,
      finalCalled: false,
      needDrain: false,
      ending: true,
      ended: true,
      finished: true,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: false,
      bufferProcessing: false,
      onwrite: [Function (anonymous)],
      writecb: null,
      writelen: 0,
      bufferedRequest: null,
      lastBufferedRequest: null,
      pendingcb: 0,
      prefinished: true,
      errorEmitted: false,
      bufferedRequestCount: 0,
      corkedRequestsFree: [CorkedRequest]
    },
    writable: false,
    allowHalfOpen: true,
    _options: { objectMode: true },
    _wrapOptions: { objectMode: true },
    _streams: [ [Deps], [Transform] ],
    length: 2,
    label: 'deps',
    [Symbol(kCapture)]: false
  }
}
```
</details>

FrederikBolding replied:
@brad-decker When using `fs` directly like this, it will be replaced by `brfs` at build time with an inlined version of the file. This is the intention for bundling preinstalled Snaps.

brad-decker replied:
@FrederikBolding sounds like from your comment that overriding the lint rule in this case is the expected outcome. In which case I'm fine resolving this entry

brad-decker reviewed (changes requested):
@Prithpal-Sooriya there's a lot of review comments that are being hidden from future reviewers by being marked resolve without any apparent resolution other than your own commentary. Could we make sure that we don't mark anything resolved that isn't actually resolved in the minds of the team please?

Prithpal-Sooriya commented:
@brad-decker thanks for the comments. I'll make sure to only resolve comments through comments before marking as resolved.

Since I can't comment on some of your responses I'll make sure to message here:

> https://github.com/MetaMask/metamask-extension/pull/23286#discussion_r1523655209
Do we not allow mixing of promises and await syntaxes?
(personal) I prefer `.then` chaining to keep sections (e.g. transformations) scoped. Likewise it allows error handling immediately than a global try-catch where you need to differentiate errors.

> https://github.com/MetaMask/metamask-extension/pull/23286#discussion_r1523662009
OK I didn't realise that this is in the full ruleset. Okay sure can add the explicit returns.

Prithpal-Sooriya committed:
chore: resolve nit comments

Add explicit returns and replace await-the syntax.
These really should be nits:
1. Explicit returns are not in the repo yet, but why is there no eslint warning for this instead of relying on catching this at review level?
2. Await-Then syntax is valid! Async/Await is just sugar coating.

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

FrederikBolding commented on yarn.lock:
```diff
@@ -4703,6 +4703,22 @@ __metadata:
+  resolution: "@metamask/message-signing-snap@npm:0.2.0"
+  dependencies:
+    "@metamask/rpc-errors": "npm:^6.1.0"
+    "@metamask/snaps-jest": "npm:^5.0.0"
```
The Snap should not be bringing in `snaps-jest`, that should be marked as a dev dependency.

Prithpal-Sooriya replied:
Odd the snap does mark `snaps-jest` as a dev dependency.
https://github.com/consensys-vertical-apps/metamask-message-signing-snap/blob/main/packages/snap/package.json#L46

Unsure how this has been dragged into a normal dep - will investigate.

Prithpal-Sooriya replied:
thanks for the heads up on this. Updated the snap dependencies 👍🏾

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
test: update snapshots

Fredrik discovered that the snapshots should not have changed,
a separate issue that needed resolving.

Prithpal-Sooriya committed:
chore: update message-signing-snap package

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
refactor: remove try-catch on pre-installed snaps

this is only ran once at build time, and there is no need currently to log if this fails. We can fatal/hard crash instead

FrederikBolding commented on app/scripts/snaps/preinstalled-snaps.ts:
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires */
+const fs = require('fs');
+
+const TEMP_PREINSTALLED_SNAPS = [];
```
Do we still need this temporary variable now that we removed the try-catch?

Prithpal-Sooriya replied:
The temp is used since the push/inserting of snaps is done within the codefence.
Can we convert a normal array into a frozen object?

Prithpal-Sooriya replied:
Update tested, that is interesting - TIL Object.freeze mutates original!

[Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBADgJwKYEtJQIYBstICYDKYGcEAXDNAmgOYDaAujALwyMDcAUJ4qutrkLFSAOjgBXCAAsAFAHIIIALZJKwuQEpuodDGRRxCMPgBq2FjADyAIwBWSYFBEAzZEgBeSGbzTQB+IhIILU4AelCYMBAAdxho1WAMMDlYCWkYKBB4ZF9MHHw1IJgkvAypVX1DY1KAN2xxJB4c-nyhILFJWTk8ECQIMFhokAQAa01OSqNTbA7peWMUKHKEGB6+spQIABoYa3FYAElVrJokWAwMgE84VSQEBGGYZdUyABFNIA)

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
refactor: remove temporary preinstalled snaps variable

Object.freeze actually mutates the original item, so we don't need a temp variable.

Prithpal-Sooriya commented on builds.yml:
```diff
@@ -19,6 +19,7 @@ buildTypes:
       - snaps
       - keyring-snaps
       - blockaid
+      - notifications
```
If we are releasing to main, we don't need codefencing.
Also we can release the initial feature under experimental - (internal state where turned off.)

danjm replied:
(I am going to leave some checkboxes to make it easy for me to track remaining work)

- [ ] Remove code fences
- [ ] add experimental toggle

Prithpal-Sooriya replied:
Thanks for this, I'll make sure to get this ticked off!

Do we have a link to some examples of experimental features. I'll do a deep dive.

danjm commented:
@Prithpal-Sooriya You asked "Are we not tracking network failures via sentry?" in relation to catching errors and returning null. Can sentry be configured to catch all network failures? Currently, sentry tracks unhandled errors, or error we expliciting capture using Sentry methods, but are you suggesting it can track all network failures as well (regardless of how the code that makes those network requests is handled)?

danjm reviewed (changes requested):
I think we are close on this PR. I noted two todos regarding code fencing, but those could be addressed in a follow up PR.

I also left a question about sentry and tracking of network request failures.  One outstanding question for me is whether we want sentry to track failures in the cases where this PR catches errors but does not rethrow them.

If not, then this PR can proceeed as is.

If yes, is it possible to configure Sentry to track all such errors? Or should we add explicit sentry capture calls in those cases? This may not be necessary, as we control the apis/servers being requests, so we should be able to track failures at some level.

danjm commented:
- [ ] Rebase needed to resolve merge conflicts

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

brad-decker commented:
Once @danjm 's comments are addressed I think I'm good to approve as well

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
refactor: requested changes - remove codefence and use experimental feature flag

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya commented on app/scripts/controllers/preferences.js:
```diff
@@ -92,6 +92,7 @@ export default class PreferencesController {
         useNativeCurrencyAsPrimaryCurrency: true,
         hideZeroBalanceTokens: false,
         petnamesEnabled: true,
+        featureNotificationsEnabled: false,
```
Similar to experimental petnames, we've placed the experimental feature notfications here in this preferences controller.

Prithpal-Sooriya commented on ui/ducks/metamask/metamask.js:
```diff
@@ -49,6 +49,7 @@ const initialState = {
     showTestNetworks: false,
     useNativeCurrencyAsPrimaryCurrency: true,
     petnamesEnabled: true,
+    featureNotificationsEnabled: false,
```
I've copied what the experimental petnames changed.

I've not been onboarded to the project, mostly winging it 🤠 

What does this `ducks` do? Is this just for the initial redux state until it uses the state inside the preferences controller?

brad-decker replied:
There is some stuff in here that is actually used and requires more scrutiny but my blanket statement for what this duck does is "confuse our contributors" :P It was at one point used as a sort of "optimistic update" aka if we wanted to change something in the background we would also change it here optimistically to update the UI early instead of waiting for the return trip from the state update from the background.

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
chore: lint fixes

Prithpal-Sooriya committed:
chore: update lavamoat policies

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya commented:
> @Prithpal-Sooriya You asked "Are we not tracking network failures via sentry?" in relation to catching errors and returning null. Can sentry be configured to catch all network failures? Currently, sentry tracks unhandled errors, or error we expliciting capture using Sentry methods, but are you suggesting it can track all network failures as well (regardless of how the code that makes those network requests is handled)?

@danjm
Ignore this - I thought sentry could track network states, but I must have misjudged. Maybe you could intercept all requests and log them, but after some Googling I don't see it supporting network requests or failures.

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

danjm commented:
@Prithpal-Sooriya I think this PR is almost good to go, except for one thing. The feature toggle added here doesn't do anything yet, so we should make sure it is not shown by default. I made a PR that achieves this https://github.com/MetaMask/metamask-extension/pull/23753

Then you can just build like this `NOTIFICATIONS=1 yarn start` or `NOTIFICATIONS=1 yarn dist`. We can remove this environment variable as soon as the toggle has a functional impact. (I meant to tell you this when I suggested removing code fences, but forgot until I read the code here)

If you are okay with that you can merge my PR into here, and then we can approve and merge this one.

danjm committed:
Move experimental toggle behind a feature flag (#23753)

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
test: fix tests

Prithpal-Sooriya committed:
fix: linting

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into NOTIFY-414-auth-hydra-flow

brad-decker commented:
@MetaMask/supply-chain needs to approve before merge

matteoscurati reviewed (approved):
LGTM ☄️

Mrtenz commented on app/scripts/controllers/authentication/authentication-controller.ts:
```diff
@@ -0,0 +1,266 @@
+  login,
+} from './services';
+
+const THIRTY_MIN_MS = 1000 * 60 * 30;
```
`@metamask/utils` has a nice utility function for this. Something like:

```
inMilliseconds(30, Duration.Minute);
```

Prithpal-Sooriya replied:
I will promise to do this in the next PR - been itching to get this merged 😛
